templates/PlayFab_InstanceApi.cpp.ejs

```diff
@@ -140,39 +158,21 @@ namespace PlayFab
     // Private PlayFabClientInstanceAPI specific
     bool PlayFabClientInstanceAPI::IsClientLoggedIn()
     {
-        return !this->GetOrCreateAuthenticationContext()->clientSessionTicket.empty();
+        return !this->m_context->clientSessionTicket.empty();
     }
 
-    void PlayFabClientInstanceAPI::MultiStepClientLogin(bool needsAttribution)
+    void PlayFabClientInstanceAPI::MultiStepClientLogin(std::shared_ptr<PlayFabAuthenticationContext> context, bool needsAttribution)
     {
-        std::shared_ptr<PlayFab::PlayFabApiSettings> apiSettings = this->GetSettings();
-        if (apiSettings == nullptr)
+        if (needsAttribution && !context->disableAdvertising && context->advertisingIdType.length() > 0 && context->advertisingIdValue.length() > 0)
         {
-            if (needsAttribution && !PlayFabSettings::disableAdvertising && PlayFabSettings::advertisingIdType.length() > 0 && PlayFabSettings::advertisingIdValue.length() > 0)
-            {
-                AttributeInstallRequest request;
-                if (PlayFabSettings::advertisingIdType == PlayFabSettings::AD_TYPE_IDFA)
-                    request.Idfa = PlayFabSettings::advertisingIdValue;
-                else if (PlayFabSettings::advertisingIdType == PlayFabSettings::AD_TYPE_ANDROID_ID)
-                    request.Adid = PlayFabSettings::advertisingIdValue;
-                else
-                    return;
-                AttributeInstall(request, nullptr, nullptr);
-            }
-        }
-        else
-        {
-            if (needsAttribution && !apiSettings->disableAdvertising && apiSettings->advertisingIdType.length() > 0 && apiSettings->advertisingIdValue.length() > 0)
-            {
-                AttributeInstallRequest request;
-                if (apiSettings->advertisingIdType == PlayFabSettings::AD_TYPE_IDFA)
-                    request.Idfa = apiSettings->advertisingIdValue;
-                else if (apiSettings->advertisingIdType == PlayFabSettings::AD_TYPE_ANDROID_ID)
-                    request.Adid = apiSettings->advertisingIdValue;
-                else
-                    return;
-                AttributeInstall(request, nullptr, nullptr);
-            }
+            AttributeInstallRequest request;
+            if (context->advertisingIdType == PlayFabSettings::AD_TYPE_IDFA)
+                request.Idfa = context->advertisingIdValue;
+            else if (context->advertisingIdType == PlayFabSettings::AD_TYPE_ANDROID_ID)
+                request.Adid = context->advertisingIdValue;
+            else
+                return;
```

Comment:
redundant else and return ?

Reply:
False. That else condition is the fail-out.
If we don't have data to send, we won't call AttributeInstall on the next line.